Provide PHP 8.1 support

weierophinney opened the pull request:
This patch provides PHP 8.1 support, via the following:

- It **DROPS** support for PHP 7.3 versions.
- It adds `~8.1.0` to the list of allowed PHP versions.
- It bumpes phpbench to the 1.1 series.
- It bumps PHPUnit to the 9.5 series.
- It bumps laminas-coding-standard to the 2.2 series.
- It bumps vfsstream to the 1.6.10alpha series.
- It adds shims for the `ClassReflection`, `MethodReflection`, and `ParameterReflection` classes from laminas-code in order to add `#[ReturnTypeWillVary]` attributes on methods extending internals, and thus allow unit tests to run.

Note: this may mean that laminas-servicemanager will not work out-of-the-box when error reporting is set to `E_ALL` if you are using the `ReflectionBasedAbstractFactory` and/or the `bin/generate-factory-for-class` and `bin/generate-deps-for-config-factory` tools, at least until laminas-code has a release that provides PHP 8.1 support.

weierophinney committed:
feat: prepare for PHP 8.1 support

- Adds `~8.1.0` to list of PHP constraints
- Bumps phpbench to 1.1 series
- Bumps PHPUnit to 9.5.5
- Bumps laminas-coding-standard to 2.2.1
- Ignore platform PHP requirements when using CI on PHP 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update baseline file

Dependencies update flagged two more Mixed value coercions; however, there were no code changes, and they are valid for the context.
As such, they have been added to the baseline.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: bump vfsstream to 1.6.10alpha series

Experimental support for PHP 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: allow testing against PHP 8.1

Imports each of `ClassReflection`, `MethodReflection`, and `ParameterReflection` from laminas-code in order to add `#[ReturnTypeWillChange]` attributes to methods that extend PHP internals, allowing tests to run.

These shims will need to be removed at a later date, once laminas-code has a version that works against PHP 8.1.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on test/TestAsset/laminas-code/ClassReflection.php:
```diff
@@ -0,0 +1,225 @@
+use function implode;
+use function strstr;
+
+class ClassReflection extends ReflectionClass implements ReflectionInterface
```
Should be documented as temporary shim

Ocramius commented on test/TestAsset/laminas-code/MethodReflection.php:
```diff
@@ -0,0 +1,460 @@
+
+use const FILE_IGNORE_NEW_LINES;
+
+class MethodReflection extends PhpReflectionMethod implements ReflectionInterface
```
Should be documented as temporary shim

Ocramius commented on test/TestAsset/laminas-code/ParameterReflection.php:
```diff
@@ -0,0 +1,123 @@
+
+use function method_exists;
+
+class ParameterReflection extends ReflectionParameter implements ReflectionInterface
```
Should be documented as temporary shim

Ocramius commented on test/autoload.php:
```diff
@@ -0,0 +1,11 @@
+    require __DIR__ . '/TestAsset/laminas-code/MethodReflection.php';
+    require __DIR__ . '/TestAsset/laminas-code/ClassReflection.php';
+
+    class_alias(\Laminas\Code\Reflection\ParameterReflection::class, \Zend\Code\Reflection\ParameterReflection::class);
```
Are these needed at all? In theory, we don't use the `Zend\` aliases at all

weierophinney replied:
Yes - when running 8.1 against lowest, there's some oddities that happen because the ocramius proxy-manager version is sufficiently low in order to allow usage with 7.3 in this library.

Ocramius replied:
do we need usage with 7.3? I checked, and that support was for proxymanager from an ancient past...

weierophinney replied:
And, as noted, this is only for testing.

weierophinney replied:
We could potentially bump it to 7.4 as the minimum. I've not done that with any of the other packages we've done 8.1 releases for, though. :slightly_frowning_face:

weierophinney replied:
Bumped PHP minimum to 7.4. :muscle:

weierophinney committed:
qa: mark PHP 8.1 shims as internal, and add todo to remove them

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: omit test shims from CS checks

- Adds exclusion to prevent CS checks on shims
- Excludes ReferenceViaFullyQualifiedName check in test autoloader

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: exclude PHP 8.1 shims from psalm checks

As they are temporary, and not part of this package

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: bump minimum supported PHP version

Bumps minimum supported PHP version to 7.4.

Bumps ocramius/proxy-manager version to 2.11 to eliminate the need for "Zend" aliases.

Removes class aliasing in test autoloader, as it is no longer necessary.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: resolve errors flagged by Psalm after move to PHP 7.4 minimum

- Add more specific types to `LazyServiceFactory` `$servicesMap` property (`array<string, class-string>`)
- Suppress an error when testing, as it's irrelevant.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

boesing commented:
> It DROPS support for PHP 7.4 versions.

I guess thats a typo?

weierophinney commented:
> > It DROPS support for PHP 7.4 versions.
> 
> I guess thats a typo?

Ooops! Updated!

Ocramius reviewed (approved):
LGTM! @boesing I would like to also merge #98, but IMO best to do an isolated release with only PHP 8.1 support first
